exercises/exercise-13/database.ts

```diff
@@ -1,13 +1,85 @@
-export class Database<T> {
+import * as fs from 'fs';
+
+type JsonScalar = boolean | number | string;
+
+export type FieldOp = |
+    {$eq: JsonScalar} |
+    {$gt: JsonScalar} |
+    {$lt: JsonScalar} |
+    {$in: JsonScalar[]};
+
+export type Query<T extends object> = |
+    {$and: Query<T>[]} |
+    {$or: Query<T>[]} |
+    {$text: string} |
+    (
+        {[field in QueryableKeys<T>]?: FieldOp}
```

Comment:
The type can be improved by using distributive conditionals:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types

This can be used to disallow `{}` being passed as query. Ideally, it would also disallow passing multiple fields in one query, but TypeScript is not yet powerful enough to detect this.

Reply:
https://github.com/fabb/typescript-exercises/blob/my_solutions/exercises/exercise-13/database.ts#L74